src/core/action.ts

```diff
@@ -79,20 +79,11 @@ export class Action<T> extends ScheduledTrigger<T> implements Read<T> {
       tag = tag.getLaterTag(delay);
 
       if (this.action.origin === Origin.physical) {
-        // If the resulting timestamp from delay is less than the current physical time
-        // on the platform, then the timestamp becomes the current physical time.
-        // Otherwise the tag is computed like a logical action's tag.
-
-        const physicalTime = getCurrentPhysicalTime();
-        if (tag.time.isEarlierThan(physicalTime)) {
-          tag = new Tag(getCurrentPhysicalTime(), 0);
-        } else {
-          tag = tag.getMicroStepsLater(1);
-        }
+        tag = new Tag(getCurrentPhysicalTime(), 0).getLaterTag(delay);
```

Comment:
I changed physical actions to be scheduled with the current physical time. Let me explain why this is more reasonable. 
First, the delay is added to the logical tag instead of the physical time. The test [DistributedCountPhysicalAfterDelay.lf](https://github.com/lf-lang/lingua-franca/blob/ts-cyclic-dependencies/test/TypeScript/src/federated/DistributedCountPhysicalAfterDelay.lf) is failed because of that. In the test, the reactor `Count` sends a message via the physical connection at the logical tag (200 msec, 0). The physical connection has a delay of `400 msec`. The test specifies that the destination should schedule the message at a time larger than `600 msec`. Nonetheless, in the previous operation, it was scheduled at the tag (400 msec, 1). Because the 'current logical time (0 sec) + delay (400 msec)' = `400 msec` is larger than the current physical time.
Second, `reactor-c` also schedules physical actions like this. ([`lf_schedule`](https://github.com/lf-lang/reactor-c/blob/ccf230f3d1027ee39987066b9e1f3211560e9f6e/core/reactor_common.c#L939-L941)).

@lhstrh Do you think this is ok?

Reply:
Mark this as resolved as @edwardalee confirmed the concept of this change.

Reply:
Hm, I'm not sure that the behavior you implemented is the behavior we want, even though it appears to match what happens in the C implementation. I'm actually having a hard time coming up with a use case for adding a logical time delay to a physical connection. @edwardalee, do we have one? What's currently implemented in `master` of `reactor-ts` is more like a "best effort" logical connection, which I can imagine being useful.

Reply:
I'm not sure about this.  When lf_schedule gets called, what is the meaning of "current logical time"?  I think the right approach here is to provide documentation, I.e., a specification.  We need a clear definition, and then we need to make sure the tests check that the definition is what we have implemented.

So, can you explain exactly what an after delay on a physical connection means?

Reply:
I used 'an after delay on a physical connection' to explain the delay of the expression `a.out ~> b.in after delay`.

Reply:
> I'm not sure about this. When lf_schedule gets called, what is the meaning of "current logical time"? I think the right approach here is to provide documentation, I.e., a specification. We need a clear definition, and then we need to make sure the tests check that the definition is what we have implemented.

I think that our stake in the ground is that federation should be semantics preserving. 

In an unfederated context, the tag would be based on the current physical time, but when the event travels from one federate to the other, there are two physical times to consider: the one at the time of sending, `Ts`, and the one at the time of receiving, `Tr`. It seems natural to add the delay to the time of _sending_, but then there is a corner case to handle: where the delay incurred by transport + clock skew is greater than the delay on the connection, which would result in the time being less the current physical time at the receiver, which in which case the receiver might already be ahead.

So perhaps the tag at the receiver should be `(max(Ts + d, Tr), 0)`? It could also be `(Tr+d, 0)`, as @byeong-gil implemented, but it seems strange to add a delay to an event originating from a physical action but then ignore the physical time that elapses while the event is in transit. Note that the `(max(Ts + d, Tr), 0)` case also ensure that the tag of the event always has a time value greater or equal to `Ts`. In the `(Tr+d, 0)` case, when `Tr < Ts` due to clock skew, we could get a reaction to an event at a time earlier than when the event was produced (if the clock skew is greater than the network latency). This, too, seems strange, and is probably the strongest reason I can come up with to not opt for that solution.

That being said, do we have any practical use for physical connections with delays? Even in unfederated programs?

Reply:
I have questions about `((max(Ts + d, Tr), 0)`. For explanation, let's say that `tc` is a clock skew and `tn` is a network latency. Also, `Tr` is the physical time of the receiver at the time the sender sends the message. Thus, `Ts = Tr + tc`. In this case, we have to select between `((max(Ts + d, Tr + tn), 0)` and `(Tr + tn + d, 0)`.

![Screenshot_20230806_134741_Samsung Notes](https://github.com/lf-lang/reactor-ts/assets/78055940/a7ec7c3d-92d3-4bed-9ee1-9ca0a134f796)

**First**, I couldn't imagine the clear specification of the delay `d`. When we use `(Tr + tn + d, 0)`, the 'actual delay' (the delay between the message sending and the action scheduling) is apparently `tn + d`, which is 'network delay + additional delay'. [HelloWorldPhysicalAfter.lf#L7-L8](https://github.com/lf-lang/playground-lingua-franca/blob/3af88a78c2576834c1204ae2c25695f0c9088ce7/examples/C/src/distributed/HelloWorldPhysicalAfter.lf#L7-L8) also describes like that. So we can use `d` when we need additional delay among the network delay. However, when we use `((max(Ts + d, Tr + tn), 0)`, the 'actual delay' becomes `max((tc + d), tn)`. What would be the definition of `d`? 


**Second**, in the https://www.lf-lang.org/docs/handbook/composing-reactors?target=lf-c#physical-connections, 
>Physical connections are useful sometimes in [Distributed-Execution](https://www.lf-lang.org/docs/handbook/distributed-execution) in situations where the nondeterministic logical delay is tolerable.

And I think that the corner case you described below can be included in `tolerable nondeterministic logical delay`, doesn't it? Because, in my view, normally the network latency is greater than the clock skew. Please fix me if I'm wrong.
>In the (Tr+d, 0) case, when Tr < Ts due to clock skew, we could get a reaction to an event at a time earlier than when the event was produced (if the clock skew is greater than the network latency).


**Lastly**, How can the receiver know the `Ts`? I'm curious about this because in https://www.lf-lang.org/docs/handbook/composing-reactors?target=lf-c#physical-connections, the message, 
>Such connections are more efficient because timestamps need not be transmitted

Reply:
To clarify, I meant for `Tr` to denote the physical time at which the event is _received_. There is no mechanism to find out what the time at the receiver is at the moment that it is sent something by another federate. Also, note that `d` and `tn` are always positive, but `tc` can be negative. My point was that the arrow you drew in the sequence diagram (where I assume the y-axis denotes physical time as recorded at each federate), could point upwards if `tn + tc < d` (assuming `Tr = Ts + tn + tc`), meaning that the event is seen by the receiver at a tag earlier than the tag at which it was produced by the sender. I did not write the documentation that you cite, but it confirms this observation.

In order to implement `(max(Ts + d, Tr), 0)`, you need to know `Ts` (`d` is just the after delay), meaning it would have to be sent along with the message, and I agree with you that this might defeat the purpose of physical connections. I don't question the utility of physical connections, but I am unsure of the utility of _delays_ along physical connections. If there is tolerance for nondeterminism in the first place, then why use a logical delay at all?

Certainly, if we don't have good use cases for it, I won't advocate implementing something complicated like `(max(Ts + d, Tr), 0)` and would stick with the simple `(Tr+d, 0)` solution. @edwardalee, are you aware of any use cases for delayed physical connections?

Reply:
I have no use case and would stick with the simple solution.